### ISSUE TITLE
Closure parameter list

### ISSUE THREAD

**Issue by dtolnay** [rust-analyzer/ungrammar#8]
The ClosureExpr node is currently defined in a way that would imply parentheses around the parameters `() body`, rather than vertical pipes `|| body`.

https://github.com/rust-analyzer/ungrammar/blob/07d5806c9345e43c5fb2baa427bc149c8808a10f/rust.ungram#L423-L425

https://github.com/rust-analyzer/ungrammar/blob/07d5806c9345e43c5fb2baa427bc149c8808a10f/rust.ungram#L135-L139

**Comment by matklad** [comment]
Yeah, I’ve noticed that as well. I am debating between just allowing | in Params, or using separate node types for closures.

I don‘t have a super great intuition for where to reuse nodes and where to use more refined types, but I lean towards reusing.

**Comment by dtolnay** [comment]
In Syn we use [`Signature`](https://docs.rs/syn/1.0.38/syn/struct.Signature.html) for all instances of function signatures involving `fn` and a different representation for [`ExprClosure`](https://docs.rs/syn/1.0.38/syn/struct.ExprClosure.html), since the argument grammar is pretty different anyway beyond just the delimiters. For example you can always get the Type of a fn parameter from syntax, but not of a closure parameter.

**Comment by matklad** [comment]
> For example you can always get the Type of a fn parameter from syntax, but not of a closure parameter.

Which is slightly different for rust-analyzer, where everything is optional. In `fn foo(x: ) { }` there will be a Param node in the CST. So, any processing of params syntax needs to deal with missing types *anyway*, so there‘s less incentive to try to lift „is parameter optional“ to the type level.
